### ISSUE TITLE
Bleadperl v5.25.3-232-g10f9b9b breaks LEONT/Const-Fast-0.014.tar.gz

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#15522]
Migrated from [rt.perl.org#128966](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=128966) (status was 'open')

Searchable as RT128966$

**Comment by p5pRT** [comment]
### From @andk

bisect


---
commit 10f9b9bf77ed09ee67da058d1ba1658a4ce67626
Author&#8203;: Yves Orton \<demerphq@&#8203;gmail\.com>
Date&#8203;:   Mon Aug 8 18&#8203;:53&#8203;:20 2016 \+0200

&nbsp;    move Internals&#8203;::hv\_clear\_placeholders\(\) to Hash&#8203;::Util&#8203;::\_clear\_placeholders\(\)

diagnostics


---
http&#8203;://www\.cpantesters\.org/cpan/report/40ee7106\-62a0\-11e6\-b236\-c19558b9f28c

perl \-V


---
Summary of my perl5 \(revision 5 version 25 subversion 4\) configuration&#8203;:
&nbsp;  Commit id&#8203;: 2e66fe905dd7ad04f1d82df888fa62650d3d2165
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux
&nbsp;    osvers=4\.3\.0\-1\-amd64
&nbsp;    archname=x86\_64\-linux
&nbsp;    uname='linux k83 4\.3\.0\-1\-amd64 \#1 smp debian 4\.3\.3\-7 \(2016\-01\-19\) x86\_64 gnulinux '
&nbsp;    config\_args='\-Dprefix=/home/sand/src/perl/repoperls/installed\-perls/perl/v5\.25\.3\-245\-g2e66fe9/79cc \-Dmyhostname=k83 \-Dinstallusrbinperl=n \-Uversiononly \-Dusedevel \-des \-Ui\_db \-Dlibswanted=cl pthread socket inet nsl gdbm dbm malloc dl ld sun m crypt sec util c cposix posix ucb BSD gdbm\_compat \-Uuseithreads \-Uuselongdouble \-DDEBUGGING=\-g'
&nbsp;    hint=recommended
&nbsp;    useposix=true
&nbsp;    d\_sigaction=define
&nbsp;    useithreads=undef
&nbsp;    usemultiplicity=undef
&nbsp;    use64bitint=define
&nbsp;    use64bitall=define
&nbsp;    uselongdouble=undef
&nbsp;    usemymalloc=n
&nbsp;    bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'
&nbsp;    ccflags ='\-fwrapv \-fno\-strict\-aliasing \-pipe \-fstack\-protector\-strong \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64 \-D\_FORTIFY\_SOURCE=2'
&nbsp;    optimize='\-O2 \-g'
&nbsp;    cppflags='\-fwrapv \-fno\-strict\-aliasing \-pipe \-fstack\-protector\-strong \-I/usr/local/include'
&nbsp;    ccversion=''
&nbsp;    gccversion='5\.3\.1 20160121'
&nbsp;    gccosandvers=''
&nbsp;    intsize=4
&nbsp;    longsize=8
&nbsp;    ptrsize=8
&nbsp;    doublesize=8
&nbsp;    byteorder=12345678
&nbsp;    doublekind=3
&nbsp;    d\_longlong=define
&nbsp;    longlongsize=8
&nbsp;    d\_longdbl=define
&nbsp;    longdblsize=16
&nbsp;    longdblkind=3
&nbsp;    ivtype='long'
&nbsp;    ivsize=8
&nbsp;    nvtype='double'
&nbsp;    nvsize=8
&nbsp;    Off\_t='off\_t'
&nbsp;    lseeksize=8
&nbsp;    alignbytes=8
&nbsp;    prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'
&nbsp;    ldflags =' \-fstack\-protector\-strong \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /usr/lib/gcc/x86\_64\-linux\-gnu/5/include\-fixed /usr/include/x86\_64\-linux\-gnu /usr/lib /lib/x86\_64\-linux\-gnu /lib/\.\./lib /usr/lib/x86\_64\-linux\-gnu /usr/lib/\.\./lib /lib
&nbsp;    libs=\-lpthread \-lnsl \-lgdbm \-ldl \-lm \-lcrypt \-lutil \-lc \-lgdbm\_compat
&nbsp;    perllibs=\-lpthread \-lnsl \-ldl \-lm \-lcrypt \-lutil \-lc
&nbsp;    libc=libc\-2\.21\.so
&nbsp;    so=so
&nbsp;    useshrplib=false
&nbsp;    libperl=libperl\.a
&nbsp;    gnulibc\_version='2\.21'
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs
&nbsp;    dlext=so
&nbsp;    d\_dlsymun=undef
&nbsp;    ccdlflags='\-Wl\,\-E'
&nbsp;    cccdlflags='\-fPIC'
&nbsp;    lddlflags='\-shared \-O2 \-g \-L/usr/local/lib \-fstack\-protector\-strong'


Characteristics of this binary \(from libperl\)&#8203;: 
&nbsp;  Compile\-time options&#8203;:
&nbsp;    HAS\_TIMES
&nbsp;    PERLIO\_LAYERS
&nbsp;    PERL\_COPY\_ON\_WRITE
&nbsp;    PERL\_DONT\_CREATE\_GVSV
&nbsp;    PERL\_HASH\_FUNC\_ONE\_AT\_A\_TIME\_HARD
&nbsp;    PERL\_MALLOC\_WRAP
&nbsp;    PERL\_OP\_PARENT
&nbsp;    PERL\_PRESERVE\_IVUV
&nbsp;    PERL\_USE\_DEVEL
&nbsp;    USE\_64\_BIT\_ALL
&nbsp;    USE\_64\_BIT\_INT
&nbsp;    USE\_LARGE\_FILES
&nbsp;    USE\_LOCALE
&nbsp;    USE\_LOCALE\_COLLATE
&nbsp;    USE\_LOCALE\_CTYPE
&nbsp;    USE\_LOCALE\_NUMERIC
&nbsp;    USE\_LOCALE\_TIME
&nbsp;    USE\_PERLIO
&nbsp;    USE\_PERL\_ATOF
&nbsp;  Built under linux
&nbsp;  Compiled at Aug 15 2016 04&#8203;:11&#8203;:41
&nbsp;  %ENV&#8203;:
&nbsp;    PERL5LIB=""
&nbsp;    PERL5OPT=""
&nbsp;    PERL5\_CPANPLUS\_IS\_RUNNING="29355"
&nbsp;    PERL5\_CPAN\_IS\_RUNNING="29355"
&nbsp;    PERL\_CANARY\_STABILITY\_NOPROMPT="1"
&nbsp;    PERL\_MM\_USE\_DEFAULT="1"
&nbsp;  @&#8203;INC&#8203;:
&nbsp;    /home/sand/src/perl/repoperls/installed\-perls/perl/v5\.25\.3\-245\-g2e66fe9/79cc/lib/site\_perl/5\.25\.4/x86\_64\-linux
&nbsp;    /home/sand/src/perl/repoperls/installed\-perls/perl/v5\.25\.3\-245\-g2e66fe9/79cc/lib/site\_perl/5\.25\.4
&nbsp;    /home/sand/src/perl/repoperls/installed\-perls/perl/v5\.25\.3\-245\-g2e66fe9/79cc/lib/5\.25\.4/x86\_64\-linux
&nbsp;    /home/sand/src/perl/repoperls/installed\-perls/perl/v5\.25\.3\-245\-g2e66fe9/79cc/lib/5\.25\.4
&nbsp;    \.
\-\- 
andreas

**Comment by p5pRT** [comment]
### From @eserte



Also affected&#8203;: BMORROW/Readonly\-Tiny\-3\.tar\.gz
See https://rt.cpan.org/Ticket/Display.html?id=117170

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @Leont

On Tue\, Aug 16\, 2016 at 9&#8203;:44 PM\, Andreas J\. Koenig via RT \<
perlbug\-followup@&#8203;perl\.org> wrote&#8203;:

> bisect
> \-\-\-\-\-\-
> commit 10f9b9bf77ed09ee67da058d1ba1658a4ce67626
> Author&#8203;: Yves Orton \<demerphq@&#8203;gmail\.com>
> Date&#8203;:   Mon Aug 8 18&#8203;:53&#8203;:20 2016 \+0200
>
>     move Internals&#8203;::hv\_clear\_placeholders\(\) to
> Hash&#8203;::Util&#8203;::\_clear\_placeholders\(\)
>
> diagnostics
> \-\-\-\-\-\-\-\-\-\-\-
> http&#8203;://www\.cpantesters\.org/cpan/report/40ee7106\-62a0\-11e6\-
> b236\-c19558b9f28c


I think the removal from Internals&#8203;:: was a bit premature\. This function is
useful outside of Hash&#8203;::Util in one particular circumstance&#8203;: when making a
hash readonly \(or actually restricted\)\. Moving this but leaving SvREADONLY
is suboptimal\, what you really want is make\_const function that takes care
of this if it's a hash\.

Or actually\, what would be even better is having a sensible split between
readonly and restricted hashes\, but that's a different discussion\.

Leon

**Comment by p5pRT** [comment]
### From @andk

also affected&#8203;: TOBYINK/Sub\-Trigger\-Lock\-0\.002\.tar\.gz

\-\- 
andreas

**Comment by p5pRT** [comment]
### From @demerphq

On 24 August 2016 at 16&#8203;:41\, Leon Timmermans \<fawaka@&#8203;gmail\.com> wrote&#8203;:
> On Tue\, Aug 16\, 2016 at 9&#8203;:44 PM\, Andreas J\. Koenig via RT
> \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
>>
>> bisect
>> \-\-\-\-\-\-
>> commit 10f9b9bf77ed09ee67da058d1ba1658a4ce67626
>> Author&#8203;: Yves Orton \<demerphq@&#8203;gmail\.com>
>> Date&#8203;:   Mon Aug 8 18&#8203;:53&#8203;:20 2016 \+0200
>>
>>     move Internals&#8203;::hv\_clear\_placeholders\(\) to
>> Hash&#8203;::Util&#8203;::\_clear\_placeholders\(\)
>>
>> diagnostics
>> \-\-\-\-\-\-\-\-\-\-\-
>>
>> http&#8203;://www\.cpantesters\.org/cpan/report/40ee7106\-62a0\-11e6\-b236\-c19558b9f28c
>
>
> I think the removal from Internals&#8203;:: was a bit premature\.

TL;DR&#8203;: It was readded some time ago\. But if you are going to use
undocumented internals features and unroll official functionality to
reimplement or extend that functionality then IMO you get to keep the
pieces when the internals changes\.

> This function is
> useful outside of Hash&#8203;::Util in one particular circumstance&#8203;: when making a
> hash readonly \(or actually restricted\)\. Moving this but leaving SvREADONLY
> is suboptimal\, what you really want is make\_const function that takes care
> of this if it's a hash\.

I have a few issues with this\.

First off\, Internals&#8203;::hv\_clear\_placehold\(\) is only needed to get rid
of placeholders stored in an \*ALREADY\* restricted hash\.

So\, if your code is meant to mark a data structure readonly then it is
not needed\.

Second\, as far as I can tell the code in Const&#8203;::Fast overlaps with and
extends Hash&#8203;::Util\, and as the official place to expose
readonly/restricted hashes IMO you should have either used Hash&#8203;::Util\,
or patched Hash&#8203;::Util to offer a way to recursively handle other types
or to perhaps expose hv\_clear\_placeholders\(\) via Hash&#8203;::Util so you
could use it as a documented function\.

However what you have done in this module is use undocumented
internals routines to unroll the official functionality\. In theory as
undocumented internals functions we should be able to remove or change
Internals&#8203;::hv\_clear\_placehold\(\) and Internals&#8203;::SvREADONLY\(\) in any way
we choose\, like I did\, and not break any well written code\. But you
have decided to violate the encapsulation and as far as I am concerned
when you do that you get to keep both pieces\. \(Like I do in
Data&#8203;::Dump&#8203;::Streamer or Sereal when something internals changes\.\)

Third\, I think that functionality like Const&#8203;::Fast should be coded in
XS\, and be part of the core\. If its meant to be fast then it should be
XS\, and since it didles internals flags that have uncertain and
ambiguous meanings \(such as the restricted/readonly ambiguity\) it
should probably be part of core and maintained as part of core\.
Essentially you are using undocumented functions to fill a
functionalitry vacuum that core has not filled\. I think it would be
better for all if this functionality lived in core\.

Having said all that&#8203;: I re\-added hv\_clear\_placehold\(\) back to
Internals some months back\. I personally think we should mark the
Internals functions deprecated and make them throw deprecation
warnings and require people to use well defined and well documented
functions to achieve the same results\.

>
> Or actually\, what would be even better is having a sensible split between
> readonly and restricted hashes\, but that's a different discussion\.

I think we should ditch restricted hashes\, and just have what you
would call readonly hashes\. But as you say that is a different
discussion\.

Yves


\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @jkeenan

On Fri\, 26 Aug 2016 06&#8203;:15&#8203;:37 GMT\, andreas\.koenig\.7os6VVqR@&#8203;franz\.ak\.mind\.de wrote&#8203;:
> also affected&#8203;: TOBYINK/Sub\-Trigger\-Lock\-0\.002\.tar\.gz
> 

Appears to have been fixed at least as of&#8203;:

This is perl 5\, version 25\, subversion 9 \(v5\.25\.9 \(v5\.25\.8\-76\-gb1b8fb6\)\) built for x86\_64\-linux\-thread\-multi

I was able to use 'cpanm' to install the distro at this point in blead\.

Ticket not yet closable until the Const\-Fast discussion is resolved\.

\-\- 
James E Keenan \(jkeenan@&#8203;cpan\.org\)

**Comment by p5pRT** [comment]
### From @iabyn

On Sun\, Dec 25\, 2016 at 06&#8203;:51&#8203;:40PM \-0800\, James E Keenan via RT wrote&#8203;:
> On Fri\, 26 Aug 2016 06&#8203;:15&#8203;:37 GMT\, andreas\.koenig\.7os6VVqR@&#8203;franz\.ak\.mind\.de wrote&#8203;:
> > also affected&#8203;: TOBYINK/Sub\-Trigger\-Lock\-0\.002\.tar\.gz
> > 
> 
> Appears to have been fixed at least as of&#8203;:
> 
> This is perl 5\, version 25\, subversion 9 \(v5\.25\.9 \(v5\.25\.8\-76\-gb1b8fb6\)\) built for x86\_64\-linux\-thread\-multi
> 
> I was able to use 'cpanm' to install the distro at this point in blead\.
> 
> Ticket not yet closable until the Const\-Fast discussion is resolved\.

But unless I've misunderstood this thread\, this ticket is no longer a
5\.26\.0 blocker\.

\-\- 
I before E\. Except when it isn't\.

**Comment by p5pRT** [comment]
### From @jkeenan

On Mon\, 20 Mar 2017 12&#8203;:59&#8203;:21 GMT\, davem wrote&#8203;:
> On Sun\, Dec 25\, 2016 at 06&#8203;:51&#8203;:40PM \-0800\, James E Keenan via RT wrote&#8203;:
> > On Fri\, 26 Aug 2016 06&#8203;:15&#8203;:37 GMT\,
> > andreas\.koenig\.7os6VVqR@&#8203;franz\.ak\.mind\.de wrote&#8203;:
> > > also affected&#8203;: TOBYINK/Sub\-Trigger\-Lock\-0\.002\.tar\.gz
> > >
> >
> > Appears to have been fixed at least as of&#8203;:
> >
> > This is perl 5\, version 25\, subversion 9 \(v5\.25\.9 \(v5\.25\.8\-76\-
> > gb1b8fb6\)\) built for x86\_64\-linux\-thread\-multi
> >
> > I was able to use 'cpanm' to install the distro at this point in
> > blead\.
> >
> > Ticket not yet closable until the Const\-Fast discussion is resolved\.
> 
> But unless I've misunderstood this thread\, this ticket is no longer a
> 5\.26\.0 blocker\.

Agreed; removing from 5\.26\.0 blockers list\.

\-\- 
James E Keenan \(jkeenan@&#8203;cpan\.org\)

**Comment by toddr** [comment]
> ### From @jkeenan
> Ticket not yet closable until the Const-Fast discussion is resolved.

@Leont is this closeable or is further discussion/decisioning needed?

**Comment by Leont** [comment]
I think it's closable, though at some point we may need to have a policy around `Internals::`
